File: Cliente50.py
import socket
import threading
import numpy as np

class TCPClient50:
    def __init__(self, ip, message_received_callback):
        self.SERVERIP = ip
        self.SERVERPORT = 4444
        self.mMessageListener = message_received_callback
        self.mRun = False
        self.out = None
        self.incoming = None

    def send_message(self, message):
        if self.out is not None and not self.out.closed:
            self.out.write(message + "\n")
            self.out.flush()

    def stop_client(self):
        self.mRun = False

    def run(self):
        self.mRun = True
        try:
            print("TCP Client" + "C: Conectando...")
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.connect((self.SERVERIP, self.SERVERPORT))
                self.out = sock.makefile(mode='w')
                print("TCP Client" + "C: Sent.")
                print("TCP Client" + "C: Done.")
                self.incoming = sock.makefile(mode='r')
                while self.mRun:
                    server_message = self.incoming.readline()
                    if server_message and self.mMessageListener:
                        self.mMessageListener(server_message.rstrip("\n"))
        except Exception as e:
            print("TCP" + "C: Error", e)


class Cliente50:
    sum = [None] *40
    def __init__(self):
        self.mTcpClient = None
        self.sc = None

    def main(self):
        objcli = Cliente50()
        objcli.iniciar()

    def iniciar(self):
        def on_message_received(message):
            self.cliente_recibe(message)

        def tcp_client_thread():
            self.mTcpClient = TCPClient50("127.0.0.1", on_message_received)
            self.mTcpClient.run()

        # Iniciar el hilo del cliente TCP
        thread = threading.Thread(target=tcp_client_thread)
        thread.start()

        salir = "n"
        print("Cliente bandera 01")
        while salir != "s":
            salir = input()
            self.cliente_envia(salir)
        print("Cliente bandera 02")

    def cliente_recibe(self, llego):
        print("CLINTE50 El mensaje::" + llego)
        if "evalua" in llego:
            arrayString = llego.split()
            funcion = arrayString[2]
            n = int(arrayString[5])
            min = int(arrayString[6])
            max = int(arrayString[7])
            dif = (float(arrayString[4]) - float(arrayString[3]))/float(n)
            print("funcion: ", funcion, "n: ", n, " min: ", min, " max: ",max, " dif: ",dif)
            self.procesar(min,max,funcion,n, dif)

    def cliente_envia(self, envia):
        if self.mTcpClient is not None:
            self.mTcpClient.send_message(envia)
    
    class EvaluadorPolinomios:
        def __init__(self, polinomio):
            self.terminos = polinomio.split("+")

        def evaluarPolinomio(self, x):
            resultado = 0
            for termino in self.terminos:
                partes = termino.strip().split("^")
                coeficiente = float(partes[0].replace("x", "").strip())
                exponente = int(partes[1].strip())
                resultado += coeficiente * x ** exponente
            return resultado

    def procesar(self, a, b, funcion, n,dif):
        poli = self.EvaluadorPolinomios(funcion)
        N = b - a
        H = 4
        d = N / H

        todos = [None] * 40
        for i in range(H - 1):
            todos[i] = self.tarea0101(i * d + a, i * d + d + a, i, poli, n,dif)
            todos[i].start()

        todos[H - 1] = self.tarea0101(d * (H - 1) + a, b, H - 1, poli, n,dif)
        todos[H - 1].start()
        
        for i in range(H):
            todos[i].join()

        sumatotal = sum(self.sum[:H])
        print("SUMA TOTAL____:", sumatotal)
        self.cliente_envia("rpta " + str(sumatotal))


    class tarea0101(threading.Thread):
        def __init__(self, min_, max_, id_, poli_, n_,dif_):
            threading.Thread.__init__(self)
            self.max = max_
            self.min = min_
            self.id = id_
            self.poli = poli_
            self.n = n_
            self.dif = dif_

        def run(self):
            integral = 0
            print(self.min, self.max)
            for i in np.arange(0,(self.max - self.min)/self.dif,1.0):
                XD = self.min + i*self.dif
                integral += self.poli.evaluarPolinomio(XD) * self.dif
                print("Hilo: ",self.id," Entrada: ",XD,"Salida: ",self.poli.evaluarPolinomio(XD)," n: ",self.n,"max",self.max,"min",self.min,"dif",self.dif)
            Cliente50.sum[self.id] = integral
            #print(" min:", self.min, " max:", self.max, " id:", self.id, " integral:", integral)
            #print("\n")
    class OnMessageReceived:
        def messageReceived(self, message):
            pass
    
    

    def tcp_client_thread(self):
        self.mTcpClient = TCPClient50("127.0.0.1", self.OnMessageReceived())
        self.mTcpClient.run()


# Llamada al método main
if __name__ == "__main__":
    objcli = Cliente50()
    objcli.iniciar()

File: Cliente50.java
package sumaNred;

import java.util.Scanner;
import java.util.*;
import sumaNred.TCPClient50;

class Cliente50 {
    public double sum[] = new double[40];
    TCPClient50 mTcpClient;
    Scanner sc;

    public static void main(String[] args) {
        Cliente50 objcli = new Cliente50();
        objcli.iniciar();
    }

    void iniciar() {
        new Thread(
                new Runnable() {
                    @Override
                    public void run() {
                        mTcpClient = new TCPClient50("127.0.0.1",
                                new TCPClient50.OnMessageReceived() {
                                    @Override
                                    public void messageReceived(String message) {
                                        ClienteRecibe(message);
                                    }
                                });
                        mTcpClient.run();
                    }
                }).start();

        String salir = "n";
        sc = new Scanner(System.in);
        // System.out.println("Cliente bandera 01");
        while (!salir.equals("s")) {
            salir = sc.nextLine();
            ClienteEnvia(salir);
        }
        // System.out.println("Cliente bandera 02");

    }

    void ClienteRecibe(String llego) {
        System.out.println("CLINTE50 El mensaje::" + llego);
        if (llego.trim().contains("evalua")) {
            String arrayString[] = llego.split("\\s+");
            String funcion = arrayString[2];
            int n = Integer.parseInt(arrayString[8]);
            int min = Integer.parseInt(arrayString[6]);
            int max = Integer.parseInt(arrayString[7]);
            float dif = (Float.parseFloat(arrayString[4]) - Float.parseFloat(arrayString[3])) / n;
            procesar(min, max, funcion, n, dif);
        }
    }

    void ClienteEnvia(String envia) {
        if (mTcpClient != null) {
            mTcpClient.sendMessage(envia);
        }
    }

    void procesar(float a, float b, String funcion, int n, float dif) {
        EvaluadorPolinomios poli = new EvaluadorPolinomios(funcion);
        float N = (b - a);
        int H = 2;// luego aumentar
        float d = (float) ((N) / H);
        // n = n/H;
        // INSTANCIA LAS TAREAS EN HILOS E INICIALIZA LOS HILOS
        Thread todos[] = new Thread[40];
        for (int i = 0; i < (H - 1); i++) {
            // System.out.println("a:" + (i * d + a) + "b" + (i * d + d + a) + " i" + i);
            todos[i] = new tarea0101((i * d + a), (i * d + d + a), i, poli, n, dif);
            todos[i].start();
        }
        // System.out.println("a" + ((d * (H - 1)) + a) + "b" + (b + 1) + " i" + (H -
        // 1));

        todos[H - 1] = new tarea0101(((d * (H - 1)) + a), b, (H - 1), poli, n, dif);
        todos[H - 1].start();

        for (int i = 0; i <= (H - 1); i++) {
            try {
                todos[i].join();
            } catch (InterruptedException ex) {
                System.out.println("error" + ex);
            }
        }

        double sumatotal = 0;
        for (int i = 0; i < H; i++) {
            sumatotal = sumatotal + sum[i];
        }
        System.out.println("SUMA TOTAL____:" + sumatotal);
        ClienteEnvia("rpta " + sumatotal);
    }

    public class tarea0101 extends Thread {
        public float max, min, n, dif;
        public int id;
        public EvaluadorPolinomios poli;

        tarea0101(float min_, float max_, int id_, EvaluadorPolinomios poli_, float n_, float dif_) {
            max = max_;
            min = min_;
            id = id_;
            poli = poli_;
            n = n_;
            dif = dif_;
        }

        public void run() {
            double integral = 0;

            for (int i = 0; i < (max - min) / dif; i++) {
                float XD = min + (i) * dif;
                integral = integral + poli.evaluarPolinomio(XD) * dif;
                System.out.println("Hilo: " + id + "Entrada: " + XD + "Salida: " + poli.evaluarPolinomio(XD) + "n: " + n
                        + "max: " + max + "min: " + min + "dif: " + dif);
            }
            sum[id] = integral;

        }
    }

}

class EvaluadorPolinomios {
    public List<String> terminos;

    EvaluadorPolinomios(String polinomio) {
        terminos = Arrays.asList(polinomio.split("\\+"));
    }

    public double evaluarPolinomio(double x) {
        double resultado = 0;
        for (String termino : terminos) {
            String[] partes = termino.trim().split("\\^");
            double coeficiente = Double.parseDouble(partes[0].replace("x", "").trim());
            int exponente = Integer.parseInt(partes[1].trim());
            resultado += coeficiente * Math.pow(x, exponente);
        }
        return resultado;
    }
}
